### PR TITLE
Added types for magicmirror-module

### PR DESCRIPTION
**Pull request by jalibu**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [X] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [X] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [X] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [X] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/magicmirror-module/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Type definitions for non-npm package magicmirror-module 2.16
+// Project: https://magicmirror.builders/
+// Definitions by: Jalibu <https://github.com/jalibu>
+//                 MichMich <https://github.com/MichMich>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 3.0
```
Comment (peterblazejewicz) [inline types/magicmirror-module/index.d.ts:6]:
```suggestion
// TypeScript Version: 3.7
```
3.0 is no longer enforced and tested on DT. You can skip this header completely (3.7 is the lowest tested version here now)

**Comment by peterblazejewicz** [comment]
There are types in the module itself:
https://github.com/MichMich/MagicMirror/blob/master/module-types.ts
what's the relation to that PR?

**Comment by jalibu** [comment]
> There are types in the module itself:
> https://github.com/MichMich/MagicMirror/blob/master/module-types.ts
> what's the relation to that PR?

@peterblazejewicz the types in the module are unmaintained, outdated and incomplete. They do not reflect the requirements for (external) thirdparty modules.

**Comment by peterblazejewicz** [comment]
thx, how is this gonna be used? as a module or global script?
https://docs.magicmirror.builders/development/introduction.html
is this expected usage? (no imports ESM or CJS, just globally scoped symbols)

File: types/magicmirror-module/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// Type definitions for non-npm package magicmirror-module 2.16
+// Project: https://magicmirror.builders/
+// Definitions by: Jalibu <https://github.com/jalibu>
+//                 MichMich <https://github.com/MichMich>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 3.7
+
+declare namespace Module {
+    /* tslint:disable:no-unnecessary-generics */
+    function register<T>(moduleName: string, moduleProperties: ModuleProperties<T>): void;
```
Comment (peterblazejewicz) [inline types/magicmirror-module/index.d.ts:10]:
```suggestion
    function register<T>(moduleName: string, moduleProperties?: ModuleProperties<T>): void;
```
docs explicitely shows example without options:
https://docs.magicmirror.builders/development/core-module-file.html#the-core-module-file 
which is correct?

**Comment by peterblazejewicz** [comment]
Do you need more module methods, like `loaded`, `start`, to make it usable?
https://docs.magicmirror.builders/development/core-module-file.html#subclassable-module-methods

**Comment by jalibu** [comment]
> thx, how is this gonna be used? as a module or global script?
> https://docs.magicmirror.builders/development/introduction.html
> is this expected usage? (no imports ESM or CJS, just globally scoped symbols)

Here is a typical example of usage: https://github.com/jalibu/MMM-Jast
The extensions are developed "side by side" to the platform. The Types provide the developers with all symbols which are available during runtime, when the module loads these extensions into its scope.




> Do you need more module methods, like `loaded`, `start`, to make it usable?
> https://docs.magicmirror.builders/development/core-module-file.html#subclassable-module-methods

One does not need them. But I'll add `init`, `start`, `getScripts` and `loaded` to get a besser user experience

File: types/magicmirror-module/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+declare namespace Module {
+    /* tslint:disable:no-unnecessary-generics */
+    function register<T>(
+        moduleName: string,
+        moduleProperties: ThisType<NonNullable<ModuleProperties<T>>> &
+            Partial<ModuleProperties<T>>,
+    ): void;
+
+    class ModuleProperties<T> {
```
Comment (peterblazejewicz) [inline types/magicmirror-module/index.d.ts:15]: This should be a type or interface. We should avoid creating symbols do not exist in the package (why: we've had a bad ride with webpack as a community with `types/webpack`)

File: types/magicmirror-module/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        init?: () => void;
+        start?: () => void;
+        stop?: () => void;
+        socketNotificationReceived?: (notification: string, payload: any) => void;
+        sendSocketNotification?: (notification: string, payload: any) => void;
+        [key: string]: any;
+    } & ThisType<NodeHelperModule>;
+}
+
+declare const config: {
```
Comment (peterblazejewicz) [inline types/magicmirror-module/index.d.ts:71]: this suggests there is a 'config' symbol somewhere when this is loaded, 
is that a case? If not, should this be a type or interface (a contract)

Comment (jalibu) [reply types/magicmirror-module/index.d.ts:79]: Yes, config is a symbol which is available within the runtime scope, provided by the platform

File: types/magicmirror-module/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
+// Type definitions for non-npm package magicmirror-module 2.16
+// Project: https://magicmirror.builders/
+// Definitions by: Jalibu <https://github.com/jalibu>
+//                 MichMich <https://github.com/MichMich>
```
Comment (peterblazejewicz) [inline types/magicmirror-module/index.d.ts:4]:
```suggestion
```
the last bit: this most probably requires MichMich explicit consent here, is that OK to reference him in PR?
I understand MichMich is the owner of original module implementation, but this is the same as owner of types definition (you can always give a reference in implementation code pointing to original idea, etc)

Comment (jalibu) [reply types/magicmirror-module/index.d.ts:4]: I can not mention him here, but I wrote him on Discord. Let's see what happens

**Comment by elibarzilay** [comment]
If @MichMich doesn't reply, then it would be better to remove them from the list (since its intention is for people who maintain the types here).

**Comment by jalibu** [comment]
@elibarzilay @peterblazejewicz I removed him. Perhaps he is on vacation.
I'd add him again, if he would ask for it

**Comment by jalibu** [comment]
Ready to merge

**Comment by MichMich** [comment]
Sorry for the slow response. I've got *zero* experience with typescript, so I won't contribute to this repo. Sorry!

**Pull request merged**